### ISSUE TITLE
MTS200 Lower switching differential temperature

### ISSUE THREAD

**Issue by FabianGabor** [krahabb/meross_lan#349]
Currently the MTS200 thermostat's switching differential's minimum value is 0.5 C which is way too high for my comfort. I have underfloor heating and 0.5 C difference means that the boiler will keep heating for a much longer amount of time than needed and will overshoot the target by at least 0.5 C (in reality by 0.7-0.8 C).

My workaround for this is an automation that looks like this: if the current temperature is higher than the set temperature by any amount, lower the target temperature by 0.5 C so the thermostat switches to idle state.
For example:

- (target) temperature is 22.5
- current temperature is 22.3
- thermostat turns on but would keep heating until it reaches (target) temperature of 22.5 + 0.5 (lowest amount of diff in the app settings)
- current temperature raises to 22.5 or higher
- automation gets triggered:
1. lowers the target temperature to 22.5 - 0.5 = 22, this will put the thermostat into idle state
2. waits a bit
3. raises the target temperature back to the original value 22 + 0.5 = 22.5
- now the target temperature is less or equal to current temperature, thermostat will remain in idle state

```
trigger:
  - platform: template
    value_template: >-
      {{ state_attr('climate.meross_lan','current_temperature') >=
      state_attr('climate.meross_lan', 'temperature') }}
action:
  - service: climate.set_temperature
    target:
      entity_id: climate.meross_lan
    data_template:
      temperature: >
        {{ state_attr('climate.meross_lan','temperature') | float - 0.5 }}
  - delay:
      hours: 0
      minutes: 0
      seconds: 15
      milliseconds: 0
  - service: climate.set_temperature
    target:
      entity_id: climate.meross_lan
    data_template:
      temperature: >
        {{ state_attr('climate.meross_lan','temperature') | float + 0.5 }}
```

**Comment by krahabb** [comment]
Hello @FabianGabor,
That's a nice trick and it might also be useful for people trying to use a different target temperature sensor than the internal one (since this is almost always reporting wrong temperatures expecially on MTS100-150 valves, see #334)

meross_lan uses 0.5 resolution in setting thermostat setpoints because this is what the Meross device expects but I can only directly confirm this for MTS100 valves which I'm using and testing so I'm not sure if the MTS200 would accept finer resolutions.

**Comment by FabianGabor** [comment]
@krahabb I sniffed and saved the logs when setting the sensitivity from 0.5 C to 3 C and back.

```
2023-12-14 07:34:32,612 - meross_iot.http_api - ERROR - Login API redirected to different region: https://iotx-eu.meross.com. Login will be re-attempted
2023-12-14 07:34:33,105 - meross_iot.http_api - INFO - Login successful against https://iotx-eu.meross.com
2023-12-14 07:34:41,729 - meross_iot.manager - INFO - Allocating new mqtt client for mqtt-eu-3.meross.com:443...
2023-12-14 07:34:42,100 - meross_iot.manager - INFO - Subscribed to topics, scheduling state update for already known devices.
2023-12-14 07:34:42,104 - meross_iot.manager - INFO - 

------- Triggering Manager Discovery, filter_device: [None] -------
2023-12-14 07:34:42,559 - root - INFO - Sysdata for Nappali (REDACTED): {'all': {'system': {'hardware': {'type': 'mts200b', 'subType': 'eu', 'version': '7.0.0', 'chipType': 'rtl8710cm', 'uuid': 'REDACTED', 'macAddress': 'REDACTED'}, 'firmware': {'version': '7.6.10', 'homekitVersion': '6.3', 'compileTime': '2023/10/13-10:32:54', 'encrypt': 1, 'wifiMac': 'REDACTED', 'innerIp': 'REDACTED', 'server': 'mqtt-eu-3.meross.com', 'port': 443, 'userId': REDACTED},...}}
2023-12-14 07:34:42,967 - root - INFO - Abilities for Nappali (REDACTED): {'payloadVersion': 1, 'ability': {'Appliance.Config.Key': {}, 'Appliance.Config.WifiList': {}, 'Appliance.Config.Wifi': {}, 'Appliance.Config.WifiX': {}, 'Appliance.Config.Trace': {}, 'Appliance.Config.Info': {}, 'Appliance.Config.OverTemp': {}, 'Appliance.System.All': {}, 'Appliance.System.Hardware': {}, 'Appliance.System.Firmware': {}, 'Appliance.System.Debug': {}, 'Appliance.System.Online': {}, 'Appliance.System.Time': {}, 'Appliance.System.Clock': {}, 'Appliance.System.Ability': {}, 'Appliance.System.Runtime': {}, 'Appliance.System.Report': {}, 'Appliance.System.Position': {}, 'Appliance.Mcu.Firmware': {}, 'Appliance.Control.Multiple': {'maxCmdNum': 3}, 'Appliance.Control.Bind': {}, 'Appliance.Control.Unbind': {}, 'Appliance.Control.Upgrade': {}, 'Appliance.Control.OverTemp': {}, 'Appliance.Control.PhysicalLock': {}, 'Appliance.Control.Thermostat.HoldAction': {}, 'Appliance.Control.Thermostat.Mode': {}, 'Appliance.Control.Thermostat.Calibration': {}, 'Appliance.Control.Thermostat.DeadZone': {}, 'Appliance.Control.Thermostat.Frost': {}, 'Appliance.Control.Thermostat.Overheat': {}, 'Appliance.Control.Thermostat.WindowOpened': {}, 'Appliance.Control.Thermostat.Schedule': {}, 'Appliance.Control.Screen.Brightness': {}, 'Appliance.Control.Thermostat.Sensor': {}, 'Appliance.Control.Thermostat.SummerMode': {}, 'Appliance.Control.Sensor.History': {}, 'Appliance.Digest.TriggerX': {}, 'Appliance.Digest.TimerX': {}}}}
2023-12-14 07:34:43,662 - root - INFO - Subscribing to topic: /appliance/REDACTED/subscribe
2023-12-14 07:34:43,734 - root - INFO - Subscribing to topic: /app/REDACTED-sniffer/subscribe
2023-12-14 07:34:43,807 - root - INFO - Subscribing to topic: /app/REDACTED/subscribe
2023-12-14 07:34:45,398 - root - INFO - Fake device received message: b'{"header":{"from":"/app/REDACTED/subscribe","messageId":"be6de34b81bfd4166a283b90b42e09b4","method":"GET","namespace":"Appliance.Control.Thermostat.DeadZone","payloadVersion":1,"sign":"REDACTED","timestamp":1702535682,"triggerSrc":"Android","uuid":"REDACTED"},"payload":{"deadZone":[{"channel":0}]}}' from topic /appliance/REDACTED/subscribe
2023-12-14 07:34:45,398 - root - INFO - b'{"header":{"from":"/app/REDACTED/subscribe","messageId":"be6de34b81bfd4166a283b90b42e09b4","method":"GET","namespace":"Appliance.Control.Thermostat.DeadZone","payloadVersion":1,"sign":"REDACTED","timestamp":1702535682,"triggerSrc":"Android","uuid":"REDACTED"},"payload":{"deadZone":[{"channel":0}]}}'
2023-12-14 07:34:54,822 - root - INFO - Fake device received message: b'{"header":{"from":"/app/REDACTED/subscribe","messageId":"7c859e2d4dcc764d55cadfd093259710","method":"SET","namespace":"Appliance.Control.Thermostat.DeadZone","payloadVersion":1,"sign":"REDACTED","timestamp":1702535694,"triggerSrc":"Android","uuid":"REDACTED"},"payload":{"deadZone":[{"channel":0,"value":5}]}}' from topic /appliance/REDACTED/subscribe
2023-12-14 07:34:54,822 - root - INFO - b'{"header":{"from":"/app/REDACTED/subscribe","messageId":"7c859e2d4dcc764d55cadfd093259710","method":"SET","namespace":"Appliance.Control.Thermostat.DeadZone","payloadVersion":1,"sign":"REDACTED","timestamp":1702535694,"triggerSrc":"Android","uuid":"REDACTED"},"payload":{"deadZone":[{"channel":0,"value":5}]}}'
2023-12-14 07:35:34,608 - root - INFO - USER-TOPIC (/app/3345217/subscribe) <- {'header': {'messageId': '04106b1032d8cc49d44a35f2f8d92584', 'namespace': 'Appliance.System.Online', 'timestamp': 1702535735, 'triggerSrc': 'CloudOnline', 'payloadVersion': 1, 'method': 'PUSH', 'sign': 'REDACTED', 'from': '/appliance/REDACTED/publish'}, 'payload': {'online': {'status': 1}}}
2023-12-14 07:35:42,703 - root - INFO - Fake device received message: b'{"header":{"from":"/app/REDACTED/subscribe","messageId":"835e483ba861c7b8fc0c5727e9ccb222","method":"GET","namespace":"Appliance.Control.Thermostat.DeadZone","payloadVersion":1,"sign":"REDACTED","timestamp":1702535699,"triggerSrc":"Android","uuid":"REDACTED"},"payload":{"deadZone":[{"channel":0}]}}' from topic /appliance/REDACTED/subscribe
2023-12-14 07:35:42,703 - root - INFO - b'{"header":{"from":"/app/REDACTED/subscribe","messageId":"835e483ba861c7b8fc0c5727e9ccb222","method":"GET","namespace":"Appliance.Control.Thermostat.DeadZone","payloadVersion":1,"sign":"REDACTED","timestamp":1702535699,"triggerSrc":"Android","uuid":"REDACTED"},"payload":{"deadZone":[{"channel":0}]}}'
2023-12-14 07:35:42,704 - root - INFO - Fake device received message: b'{"header":{"from":"/app/REDACTED/subscribe","messageId":"cf414aae42e1c847456d61c25c0746ae","method":"SET","namespace":"Appliance.Control.Thermostat.DeadZone","payloadVersion":1,"sign":"REDACTED","timestamp":1702535703,"triggerSrc":"Android","uuid":"REDACTED"},"payload":{"deadZone":[{"channel":0,"value":35}]}}' from topic /appliance/REDACTED/subscribe
2023-12-14 07:35:42,704 - root - INFO - b'{"header":{"from":"/app/REDACTED/subscribe","messageId":"cf414aae42e1c847456d61c25c0746ae","method":"SET","namespace":"Appliance.Control.Thermostat.DeadZone","payloadVersion":1,"sign":"REDACTED","timestamp":1702535703,"triggerSrc":"Android","uuid":"REDACTED"},"payload":{"deadZone":[{"channel":0,"value":35}]}}'
2023-12-14 07:35:42,704 - root - INFO - USER-TOPIC (/app/3345217/subscribe) <- {'header': {'messageId': '240f58dd57583d98afe7e974fce5aac9', 'namespace': 'Appliance.System.Online', 'timestamp': 1702535736, 'triggerSrc': 'CloudOnline', 'payloadVersion': 1, 'method': 'PUSH', 'sign': 'REDACTED', 'from': '/appliance/REDACTED/publish'}, 'payload': {'online': {'status': 1}}}
2023-12-14 07:35:42,704 - root - INFO - Fake device received message: b'{"header":{"from":"/app/REDACTED/subscribe","messageId":"499a2bbd0dd09215383ea0efe2dd6ff2","method":"GET","namespace":"Appliance.Control.Thermostat.DeadZone","payloadVersion":1,"sign":"REDACTED","timestamp":1702535735,"triggerSrc":"Android","uuid":"REDACTED"},"payload":{"deadZone":[{"channel":0}]}}' from topic /appliance/REDACTED/subscribe
2023-12-14 07:35:42,705 - root - INFO - b'{"header":{"from":"/app/REDACTED/subscribe","messageId":"499a2bbd0dd09215383ea0efe2dd6ff2","method":"GET","namespace":"Appliance.Control.Thermostat.DeadZone","payloadVersion":1,"sign":"REDACTED","timestamp":1702535735,"triggerSrc":"Android","uuid":"REDACTED"},"payload":{"deadZone":[{"channel":0}]}}'
2023-12-14 07:35:42,803 - root - INFO - Fake device received message: b'{"header":{"from":"/app/REDACTED/subscribe","messageId":"3aca5972b20602359244a803535eadba","method":"GET","namespace":"Appliance.Control.Thermostat.DeadZone","payloadVersion":1,"sign":"REDACTED","timestamp":1702535743,"triggerSrc":"Android","uuid":"REDACTED"},"payload":{"deadZone":[{"channel":0}]}}' from topic /appliance/REDACTED/subscribe
2023-12-14 07:35:42,803 - root - INFO - b'{"header":{"from":"/app/REDACTED/subscribe","messageId":"3aca5972b20602359244a803535eadba","method":"GET","namespace":"Appliance.Control.Thermostat.DeadZone","payloadVersion":1,"sign":"REDACTED","timestamp":1702535743,"triggerSrc":"Android","uuid":"REDACTED"},"payload":{"deadZone":[{"channel":0}]}}'
2023-12-14 07:35:55,124 - root - INFO - Fake device received message: b'{"header":{"from":"/app/REDACTED/subscribe","messageId":"fda879c866c63e28656870065a767e24","method":"SET","namespace":"Appliance.Control.Thermostat.DeadZone","payloadVersion":1,"sign":"REDACTED","timestamp":1702535755,"triggerSrc":"Android","uuid":"REDACTED"},"payload":{"deadZone":[{"channel":0,"value":5}]}}' from topic /appliance/REDACTED/subscribe
2023-12-14 07:35:55,124 - root - INFO - b'{"header":{"from":"/app/REDACTED/subscribe","messageId":"fda879c866c63e28656870065a767e24","method":"SET","namespace":"Appliance.Control.Thermostat.DeadZone","payloadVersion":1,"sign":"REDACTED","timestamp":1702535755,"triggerSrc":"Android","uuid":"REDACTED"},"payload":{"deadZone":[{"channel":0,"value":5}]}}'
2023-12-14 07:36:01,349 - root - INFO - Fake device received message: b'{"header":{"from":"/app/REDACTED/subscribe","messageId":"70ee0120d481e4f3b611eb028f5a0bdc","method":"SET","namespace":"Appliance.Control.Thermostat.DeadZone","payloadVersion":1,"sign":"REDACTED","timestamp":1702535761,"triggerSrc":"Android","uuid":"REDACTED"},"payload":{"deadZone":[{"channel":0,"value":35}]}}' from topic /appliance/REDACTED/subscribe
2023-12-14 07:36:01,349 - root - INFO - b'{"header":{"from":"/app/REDACTED/subscribe","messageId":"70ee0120d481e4f3b611eb028f5a0bdc","method":"SET","namespace":"Appliance.Control.Thermostat.DeadZone","payloadVersion":1,"sign":"REDACTED","timestamp":1702535761,"triggerSrc":"Android","uuid":"REDACTED"},"payload":{"deadZone":[{"channel":0,"value":35}]}}'
2023-12-14 07:36:01,446 - root - INFO - Fake device received message: b'{"header":{"from":"/app/REDACTED/subscribe","messageId":"0931098b2fe7c8ae3982db1b1e0d1cbf","method":"GET","namespace":"Appliance.System.Ability","payloadVersion":1,"sign":"REDACTED","timestamp":1702535761,"triggerSrc":"Android","uuid":"REDACTED"},"payload":{}}' from topic /appliance/REDACTED/subscribe
2023-12-14 07:36:01,446 - root - INFO - b'{"header":{"from":"/app/REDACTED/subscribe","messageId":"0931098b2fe7c8ae3982db1b1e0d1cbf","method":"GET","namespace":"Appliance.System.Ability","payloadVersion":1,"sign":"REDACTED","timestamp":1702535761,"triggerSrc":"Android","uuid":"REDACTED"},"payload":{}}'
2023-12-14 07:36:08,231 - meross_iot.manager - INFO - Fetch and update done
2023-12-14 07:36:08,231 - meross_iot.manager - INFO - 
------- Manager Discovery ended -------
```

Here are the commands:

```
Command Description: Setting switching differential value to 3.5 C

Request: 
{"header":{"from":"/app/<appIdValue>/subscribe","messageId":"<messageIdValue>","method":"GET","namespace":"Appliance.Control.Thermostat.DeadZone","payloadVersion":1,"sign":"<signValue>","timestamp":1702535682,"triggerSrc":"Android","uuid":"<uuidValue>"},"payload":{"deadZone":[{"channel":0}]}}
Response: 
{"deadZone": [{"channel": 0, "value": 35, "min": 5, "max": 35, "lmTime": 1702535743}]}

Command Description: Setting switching differential value to 0.5 C

Request: 
{"header":{"from":"/app/<appIdValue>/subscribe","messageId":"<messageIdValue>","method":"SET","namespace":"Appliance.Control.Thermostat.DeadZone","payloadVersion":1,"sign":"<signValue>","timestamp":1702535694,"triggerSrc":"Android","uuid":"<uuidValue>"},"payload":{"deadZone":[{"channel":0,"value":5}]}}
Response: 
{"deadZone": [{"channel": 0, "value": 5, "min": 5, "max": 35, "lmTime": 1702535755}]}
```

What I see is MQTT. `deadZone` value is the switching differential value. Sadly the min value is 5 but I'm not sure if this is a limit set by the app and the sniffer was getting that value or it's a limit set by the firmware. I'm curious what would happen if a SET request with a deadzone value of 1 would be made. Sadly I don't have mqtt set up yet for the devices.

**Comment by FabianGabor** [comment]
@krahabb I've seen a couple of mentions of deadZone in the code (haven't seen any function tied to it). Could you please add an option to change the sensitivity into a beta version so I could try it out and report back if it works? 

I wonder what could happen if the device would receive a value of 1 instead of the min 5 specified.

**Comment by krahabb** [comment]
Sure! thank you for the inspection...I'll add the 'deadZone' support into meross_lan since it is actually almost ignored

In my opinion the 0.5 resolution is 'encoded' in the device hw/fw

**Comment by FabianGabor** [comment]
@krahabb Thank you for implementing this feature. Will there be a beta to test it?

**Comment by krahabb** [comment]
@FabianGabor ,
The fix went into a 'huge' dev branch with a lot of code-base rewriting and so 'highly-experimental'
Nevertheless, now that I've published the stable `Cloudy.5`, the next beta will start containing this dev branch and will come soon...maybe today

**Comment by krahabb** [comment]
You could anyway use (with any meross_lan release tho) the meross_lan.request service and send almost any kind of payload to the device. Just set the proper device_id to target your (configured) device and fill the relevant parameters (the key parameter is not needed if the device is configured since by default that's the one used)
Latest releases are better since will show you the actual response from the device for any  request made

**Comment by FabianGabor** [comment]
@krahabb Thanks for the implementation. Sadly it seems that the min value is set by the device and won't accept lower value than 5 (0.5 C).

Tried it with:

```
service: meross_lan.request
data:
  payload: "{\"deadZone\":[{\"channel\":0,\"value\":1}]}"
  device_id: *****
  protocol: auto
  method: SET
  namespace: Appliance.Control.Thermostat.DeadZone
  host: *****
```

Afterwards I've checked the set values with a GET request. Values below 5 are not accepted. Values that are not multiples of 5 are floored (14 to 10 or 1.0 C, 16 to 15 or 1.5C). 

I've created a post on the Meross US forum titled **MTS200 lower switching differential** so anyone interested can upvote it. Sadly I don't know how to share the url from the app and it seems the forum isn't accessible from a browser (or I cannot find it by googling it).
